### PR TITLE
[Issue 7372][docs] Add detailed info for retention policy

### LINKED ISSUE #7372: Messages on topic are deleted on restart of standalone cluster
**Describe the bug**
Messages on a topic are deleted / not available anymore after standalone cluster shutdown and restart.

**To Reproduce**
Produce messages on a topic in a standalone cluster on a namespace with retention_period = -1.
Cancel the standalone cluster (Ctrl-C).
Restart the standalone cluster.
Browse pulsar manager and look at the storage of the topic. The pulsar manager reports 2 ledgers and total space used: 350Mb.
After a while (a few minutes) refresh the page.  Pulsar manager reports the topic contains 0 Mb. One of the ledgers has disappeared.

**Expected behavior**
It was expected that messages are kept forever as the retention_period on the namespace is set to -1.
**Screenshots**

**Desktop (please complete the following information):**
 - OS: IOS
 - Pulsar 2.6.0

**Additional context**

### PR DESCRIPTION
**Pull request by Jennifer88huang**
Fixes #7372 

### Motivation
The retention policy is not clearly documented.

### Modifications
Add detailed info on retention policy.
Add default value for "defaultRetentionTimeInMinutes"

### CODE REVIEW

File: site2/docs/cookbooks-retention-expiry.md [hunk 1]
```diff
@@ -27,90 +27,90 @@ Pulsar's [admin interface](admin-api-overview.md) enables you to manage both ret
...
+- If you do not want to limit the message by time, set the value of time limit to `-1`.
+- If you do not want to limit the message by size, set the value of size limit to `-1`.
+- If you need infinite retention, set the values of time limit and size limit to `-1`.
+- If you want to disable retention policy, set the values of time limit and size limit to `0`. Retention policy is disabled by default.
 
-The retention settings apply to all messages on topics that do not have any subscriptions, or if there are subscriptions, to messages that have been acked by all subscriptions. The retention policy settings do not affect unacknowledged messages on topics with subscriptions -- these are instead controlled by the backlog quota (see below).
+When you set a size limit of, for example, 10 gigabytes, and set the time limit to `-1`, then the acknowledged messages in all topics in the namespace are retained until the topic size reaches the size limit(10 gigabytes). If you set a time limit of, for example, 1 day, and set the size limit to `-1`, then the acknowledged messages for all topics in the namespace are retained for 1 day.
```
Comment (Huanli-Meng) [inline site2/docs/cookbooks-retention-expiry.md:41]:
```suggestion
When you set a size limit of, for example, 10 GB, and set the time limit to `-1`, then the acknowledged messages in all topics in the namespace are retained until the topic size reaches the size limit (10 GB). If you set a time limit of, for example, 1 day, and set the size limit to `-1`, then the acknowledged messages for all topics in the namespace are retained for 1 day.
```

File: site2/docs/cookbooks-retention-expiry.md [hunk 2]
```diff
@@ -27,90 +27,90 @@ Pulsar's [admin interface](admin-api-overview.md) enables you to manage both ret
...
+
+The retention settings apply to all messages on topics that do not have any subscriptions, or to messages that have been acked by all subscriptions. The retention policy settings do not affect unacknowledged messages on topics with subscriptions. The unacknowledged messages are controlled by the backlog quota.
```
Comment (Huanli-Meng) [inline site2/docs/cookbooks-retention-expiry.md:43]:
```suggestion
The retention settings apply to all messages on topics that do not have any subscriptions, or to messages that have been acknowledged by all subscriptions. The retention policy settings do not affect unacknowledged messages on topics with subscriptions. The unacknowledged messages are controlled by the backlog quota.
```

File: site2/docs/cookbooks-retention-expiry.md [hunk 3]
```diff
@@ -27,90 +27,90 @@ Pulsar's [admin interface](admin-api-overview.md) enables you to manage both ret
...
 
 When a retention limit is exceeded, the oldest message is marked for deletion until the set of retained messages falls within the specified limits again.
 
 ### Defaults
 
-There are two configuration parameters that you can use to set [instance](reference-terminology.md#instance)-wide defaults for message retention: [`defaultRetentionTimeInMinutes=0`](reference-configuration.md#broker-defaultRetentionTimeInMinutes) and [`defaultRetentionSizeInMB=0`](reference-configuration.md#broker-defaultRetentionSizeInMB).
+You can use the following two configuration parameters to set instance-wide defaults for message retention: `defaultRetentionTimeInMinutes` and `defaultRetentionSizeInMB`. Both parameters are set to `0` by default.
```
Comment (Huanli-Meng) [inline site2/docs/cookbooks-retention-expiry.md:49]: change instance-wide to instance-level?

File: site2/docs/cookbooks-retention-expiry.md [hunk 4]
```diff
@@ -27,90 +27,90 @@ Pulsar's [admin interface](admin-api-overview.md) enables you to manage both ret
...
-You must set **both** a *size limit* and a *time limit*. In the case where you don't want to limit by either time or set, the value must be set to `-1`. Retention policy will be effectively disabled and it won't prevent the deletion of acknowledged messages when either size or time limit is set to 0.
+You can set a retention policy for a namespace by specifying the namespace, a size limit and a time limit in `pulsar-admin`, REST API and Java.
 
-##### Examples
+<!--DOCUSAURUS_CODE_TABS-->
+<!--pulsar-admin-->
+You can use the [`set-retention`](reference-pulsar-admin.md#namespaces-set-retention) subcommand and specify a namespace, a size limit using the `-s`/`--size` flag, and a time limit using the `-t`/`--time` flag. 
 
-To set a size limit of 10 gigabytes and a time limit of 3 hours for the `my-tenant/my-ns` namespace:
+In the following example, size limit is set to 10 gigabytes and time limit is set to 3 hours for the `my-tenant/my-ns` namespace:
```
Comment (Huanli-Meng) [inline site2/docs/cookbooks-retention-expiry.md:61]:
```suggestion
In the following example, the size limit is set to 10 GB and time limit is set to 3 hours for the `my-tenant/my-ns` namespace:
```

File: site2/docs/cookbooks-retention-expiry.md [hunk 5]
```diff
@@ -27,90 +27,90 @@ Pulsar's [admin interface](admin-api-overview.md) enables you to manage both ret
...
 
 ```shell
 $ pulsar-admin namespaces set-retention my-tenant/my-ns \
   --size 10G \
   --time 3h
 ```
 
-To set retention where time limit is ignored and the size limit of 1 terabyte determines retention:
+In the following example, time is not limited and size limit is set to 1 terabyte. The size limit determines the retention result.
```
Comment (Huanli-Meng) [inline site2/docs/cookbooks-retention-expiry.md:69]:
```suggestion
In the following example, the time is not limited while the size limit is set to 1 TB. The size limit determines the retention result.
```

File: site2/docs/cookbooks-retention-expiry.md [hunk 6]
```diff
@@ -27,90 +27,90 @@ Pulsar's [admin interface](admin-api-overview.md) enables you to manage both ret
...
 
 ```shell
 $ pulsar-admin namespaces set-retention my-tenant/my-ns \
   --size 1T \
   --time -1
 ```
 
-To set retention where size limit is ignored and the time limit of 3 hours determines retention:
+In the following example, size is not limited and time limit is set to 3 hours. The time limit determines the retention result.
```
Comment (Huanli-Meng) [inline site2/docs/cookbooks-retention-expiry.md:77]:
```suggestion
In the following example, the size is not limited and the time limit is set to 3 hours. The time limit determines the retention result.
```

**Review by Huanli-Meng — COMMENTED** [review]
leave my comments. PTAL. 
BTW, the original issue seems to occur in Pulsar 2.6.0, could you please help confirm whether docs in previous releases should be updated two? Thanks.

**Review by ErikJansenIRefact — COMMENTED** [review]
Hi,

In general good improvements in describing the behaviour of the retention settings.
I would suggest however that you describe the behaviour when both settings (time and size) have a value > 0. Are the messages now removed if one of the thresholds exceeds or only if both of them are exceeded. So are the settings a "logical and" or a "logical or"?

And is the 0-value treated as a special value? So, if size = 10Gb and time=0 are message now removed if size > 10Gb?
And if size = 10Gb and time = 1d are messages now removed if either the size or the time is exceeded?

**Comment by lhotari** [comment]
> And is the 0-value treated as a special value?

@ErikJansenIRefact 
Yes, `0` is a special value. Setting either limit to `0` will disable retention. That's the catch. 
PR #8358 introduced new behavior that prevents setting just one of the values to 0 via the Admin API. To disable the retention policy, both values have to be set to 0 when using the Admin API. 
The "setting either limit to 0 will disable retention" behavior still applies to existing configurations (after upgrading to 2.7.0 when it's out) and also for `defaultRetentionTimeInMinutes` and `defaultRetentionSizeInMB` settings. I wonder how to get this documented?

**Comment by lhotari** [comment]
> I would suggest however that you describe the behaviour when both settings (time and size) have a value > 0. Are the messages now removed if one of the thresholds exceeds or only if both of them are exceeded. So are the settings a "logical and" or a "logical or"?

good points. explaining this would be useful. I believe it's described somewhere. It's a "logical or". Whatever limit is hit first will cause the data to not be retained (and thus a candidate for deletion).

**Comment by lhotari** [comment]
The code tells the truth:
https://github.com/apache/pulsar/blob/a8eaeb5eb6ddabcbc59030ea14207b43ca76237a/managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedLedgerImpl.java#L2091-L2097

The retention happens only when both conditions are met.

**Comment by Jennifer88huang** [comment]
@ErikJansenIRefact 

> Hi,
> 
> In general good improvements in describing the behaviour of the retention settings.
> I would suggest however that you describe the behaviour when both settings (time and size) have a value > 0. Are the messages now removed if one of the thresholds exceeds or only if both of them are exceeded. So are the settings a "logical and" or a "logical or"?
> 
> And is the 0-value treated as a special value? So, if size = 10Gb and time=0 are message now removed if size > 10Gb?
> And if size = 10Gb and time = 1d are messages now removed if either the size or the time is exceeded?

Good question.  it's "logical or". However, if you set either to `0`, the retention is disabled.
@lhotari Thank you for helping answer the question, and provide the related doc. 
I'll add more details here to make it clear.

**Comment by Jennifer88huang** [comment]
@ErikJansenIRefact I've listed the value settings in a table, is this way much clearer?

**Comment by Jennifer88huang** [comment]
@lhotari according to your comment,
- in 2.6.x, the setting is sth like this
image

- in 2.7, the setting is sth like this.
image

Do I grasp your meaning correctly? I've made adjustment in docs for both 2.7 and 2.6.1 accordingly (add the above table and examples). Any issue, feel free to comment.

**Comment by lhotari** [comment]
> Do I grasp your meaning correctly? I've made adjustment in docs for both 2.7 and 2.6.1 accordingly (add the above table and examples). Any issue, feel free to comment.

Yes, perfect. Good work @Jennifer88huang

**Comment by lhotari** [comment]
@Jennifer88huang  just one minor comment about "Messages are deleted when either time or size reaches the limit". The might lead to misunderstanding that deletion applies for all messages when someone looks at this table and doesn't read the other parts where it's described. I'm not sure how to refer that this only applies to the messages that have been acknowledged or don't have an active subscription at all.

**Comment by Jennifer88huang** [comment]
@lhotari great, thanks for your reminder. How about replacing "Messages are deleted when either time or size reaches the limit" with one of the followings
- "Acknowledged messages or messages with no active subscription will be deleted" 
- "Acknowledged messages or messages with no active subscription will not be retained"

**Comment by lhotari** [comment]
> "Acknowledged messages or messages with no active subscription will not be retained"

+1

**Comment by Jennifer88huang** [comment]
@lhotari thanks for your quick response. Updated, PTAL again.

**Comment by lhotari** [comment]
@Jennifer88huang I guess the "when either..." part is needed too. Something like "Acknowledged messages or messages with no active subscription will not be retained when either time or size reaches the limit."?

**Comment by Jennifer88huang** [comment]
@lhotari oops, it's my mistake. I'm sorry I forget to add the condition...
Thank you very much for your reminder.

**Review by lhotari — APPROVED** [review]

**Comment by lhotari** [comment]
I read the document once more and it seems after all that it would be better to simplify the table. One suggestion is to replace "Acknowledged messages or messages with no active subscription will not be retained when either time or size reaches the limit." in the table with "Based on time and size limits" since it would be more consistent with the other rows in the same table. 
WDYT?

**Comment by Jennifer88huang** [comment]
@lhotari Thanks for your feedback. 
I'm not sure if we use "Based on time and size limits", whether users are clear about the meaning it conveys. 
@ErikJansenIRefact Do you have any opinion on it? Which is more clear to you?

**Review by sijie — APPROVED** [review]

**Pull request merged**